File: workspace/src/arm_behavior/src/simulation_control.cpp
#include "arm_behavior/simulation_control.hpp"

#include <chrono>
#include <future>
#include <stdexcept>
#include <string>

#include "std_srvs/srv/trigger.hpp"

namespace
{

constexpr char kPauseSimulationService[] = "/physics_sync_node/pause_simulation";
constexpr char kResumeSimulationService[] = "/physics_sync_node/resume_simulation";

void call_simulation_service(
  const rclcpp::Node::SharedPtr & node,
  const std::string & service_name,
  std::chrono::seconds timeout)
{
  auto client = node->create_client<std_srvs::srv::Trigger>(service_name);
  if (!client->wait_for_service(timeout)) {
    throw std::runtime_error(
            "Simulation service '" + service_name + "' was not available after " +
            std::to_string(timeout.count()) + " seconds");
  }

  auto future = client->async_send_request(std::make_shared<std_srvs::srv::Trigger::Request>());
  if (future.wait_for(timeout) != std::future_status::ready) {
    client->remove_pending_request(future);
    throw std::runtime_error(
            "Simulation service '" + service_name + "' timed out after " +
            std::to_string(timeout.count()) + " seconds");
  }

  const auto response = future.get();
  if (!response->success) {
    throw std::runtime_error(
            response->message.empty() ?
            "Simulation service '" + service_name + "' rejected the request" :
            response->message);
  }
}

}  // namespace

namespace arm
{

void pause_simulation(const rclcpp::Node::SharedPtr & node, std::chrono::seconds timeout)
{
  call_simulation_service(node, kPauseSimulationService, timeout);
}

void resume_simulation(const rclcpp::Node::SharedPtr & node, std::chrono::seconds timeout)
{
  call_simulation_service(node, kResumeSimulationService, timeout);
}

}  // namespace arm

File: workspace/src/grasp_pose_provider/grasp_pose_provider/simulation_control.py
"""Helpers for pausing and resuming the synchronized physics loop."""

import time

import rclpy
from std_srvs.srv import Trigger


PHYSICS_SYNC_NODE_NAME = 'physics_sync_node'
PAUSE_SIMULATION_SERVICE = (
    f'/{PHYSICS_SYNC_NODE_NAME}/pause_simulation'
)
RESUME_SIMULATION_SERVICE = (
    f'/{PHYSICS_SYNC_NODE_NAME}/resume_simulation'
)


def pause_simulation(node, parameters):
    """Pause physics advancement through the physics-sync node.

    ``parameters`` is the node's
    :class:`~grasp_pose_provider.node_parameters.GraspPoseProviderParameters`;
    the service timeout comes from it.
    """
    _call_trigger_service(
        node,
        PAUSE_SIMULATION_SERVICE,
        parameters.simulation_service_timeout_sec,
    )


def resume_simulation(node, parameters):
    """Resume physics advancement through the physics-sync node."""
    _call_trigger_service(
        node,
        RESUME_SIMULATION_SERVICE,
        parameters.simulation_service_timeout_sec,
    )


def _call_trigger_service(node, service_name, timeout_sec):
    client = node.create_client(Trigger, service_name)
    try:
        if not client.wait_for_service(timeout_sec=timeout_sec):
            raise RuntimeError(
                f"Simulation service '{service_name}' not available after "
                f'{timeout_sec}s.'
            )

        future = client.call_async(Trigger.Request())
        deadline = time.monotonic() + timeout_sec
        while rclpy.ok() and not future.done() and time.monotonic() < deadline:
            time.sleep(0.01)

        response = future.result() if future.done() else None
        if response is None:
            raise RuntimeError(
                f"Simulation service '{service_name}' failed or timed out."
            )
        if not response.success:
            raise RuntimeError(
                response.message
                or f"Simulation service '{service_name}' rejected the request."
            )
    finally:
        node.destroy_client(client)
